### PR TITLE
Find the next valid folder name when owner name is not unique

### PR DESCRIPTION
**Pull request by allenwq**
Introduced a gem 'filename' here to generate sequential file names.

### CODE REVIEW

File: app/models/course/material/folder.rb [hunk 1]
```diff
@@ -21,6 +23,24 @@ def self.after_course_create(course)
...
+  # Find the next valid name whose format is based on the the name given.
+  # If the base_name has been taken, the next name will be generate in the format of: base_name(0),
+  # base_name(1), etc.
+  # @param [String] base_name The name which we want to name the folder.
+  # @return [String] The next valid name that haven't been taken.
+  def next_valid_name(base_name = name)
+    return base_name unless parent && base_name
+
+    names_taken = existing_names.map(&:downcase)
+    name_generator = FileName.new(base_name, path: :relative, add: :always,
```
Comment (lowjoel) [inline app/models/course/material/folder.rb]: I don't see why we need a gem for this... there's still n queries at the end of the day.

Comment (allenwq) [inline app/models/course/material/folder.rb]: why still n queries ? 
This gem is quite powerful for generating filenames. it can also generate numbers before the extension, like: base_name(0).txt, base_name(1).txt

Comment (lowjoel) [inline app/models/course/material/folder.rb]: because every time you generate a file name, that file name might already be taken in another thread.

Comment (allenwq) [inline app/models/course/material/folder.rb]: But for the gem, it's for generating file names only, not for reduce queries I think ?

Comment (allenwq) [inline app/models/course/material/folder.rb]: The gem is used for generating the next name in sequential order for me, then I will check the name to see if it's valid. If not I'll ask the gem to generate again.

File: app/models/course/material/folder.rb [hunk 2]
```diff
@@ -21,6 +23,24 @@ def self.after_course_create(course)
...
+                                             format: '(%d)', delimiter: '')
+
+    new_name = base_name
+    new_name = name_generator.create while names_taken.include?(new_name.downcase)
```
Comment (lowjoel) [inline app/models/course/material/folder.rb]: this approach is still susceptible to a race condition...

File: app/models/course/material/folder.rb [hunk 3]
```diff
@@ -35,4 +55,14 @@ def validate_name_is_unique_among_materials
     conflicts = parent.materials.where('lower(name) = ?', name.downcase)
     errors.add(:name, :taken) if conflicts.any?
   end
+
+  def existing_names
+    parent.children.where.not(id: self).pluck(:name) + parent.materials.pluck(:name)
+  end
+
+  def assign_valid_name
+    return unless name_changed? && owner
```
Comment (lowjoel) [inline app/models/course/material/folder.rb]: owner.present? might be clearer.

Comment (lowjoel) [inline app/models/course/material/folder.rb]: wait, why is this necessary?

Comment (lowjoel) [inline app/models/course/material/folder.rb]: If the name of the folder changed, that means someone tried to change the name of the folder directly? Why should this generate another name?

File: app/models/course/material/folder.rb [hunk 4]
```diff
@@ -35,4 +55,14 @@ def validate_name_is_unique_among_materials
     conflicts = parent.materials.where('lower(name) = ?', name.downcase)
     errors.add(:name, :taken) if conflicts.any?
   end
+
+  def existing_names
+    parent.children.where.not(id: self).pluck(:name) + parent.materials.pluck(:name)
```
Comment (lowjoel) [inline app/models/course/material/folder.rb]: this is OK, but if we have a similar query again in future we should have a method called siblings.

File: spec/models/course/assessment/category_spec.rb [hunk 5]
```diff
@@ -43,5 +43,32 @@
...
+      let(:common_title) { 'Assessments' }
+      let!(:category) { create(:course_assessment_category, course: course, title: common_title) }
+
+      context 'after category was created' do
+        subject { build(:course_assessment_category, title: common_title, course: course) }
+
+        it 'creates a folder with the proper name' do
+          subject.save
+
+          expect(subject.folder.name).to eq(subject.title + '(0)')
```
Comment (lowjoel) [inline spec/models/course/assessment/category_spec.rb]: usually there's a space before the (

File: app/models/course/material/folder.rb (hunk 3 above)
Comment (allenwq) [inline app/models/course/material/folder.rb]: This only applies to the those folders who have the owner, and the folder with owner's name is not allowed to change directly.

File: spec/models/course/assessment/category_spec.rb (hunk 5 above)
Comment (lowjoel) [inline spec/models/course/assessment/category_spec.rb]: But i don't know how this can be internationalised.

File: app/models/course/material/folder.rb (hunk 3 above)
Comment (lowjoel) [inline app/models/course/material/folder.rb]: So then why is this function necessary?

File: spec/models/course/assessment/category_spec.rb (hunk 5 above)
Comment (allenwq) [inline spec/models/course/assessment/category_spec.rb]: OK will add space then.

I think it's fine if we use numbers ? my Chinese version windows also appends strings like `'(0)'` to the files with Chinese names.

File: app/models/course/material/folder.rb (hunk 3 above)
Comment (allenwq) [inline app/models/course/material/folder.rb]: Owner will change the folder name after its own name changed right ?  so that time `name_changed?` returns `true`

Comment (lowjoel) [inline app/models/course/material/folder.rb]: then you will change the name of the folder twice.

File: app/models/course/material/folder.rb (hunk 2 above)
Comment (allenwq) [inline app/models/course/material/folder.rb]: I agree.. but it's model level, I don't think this can be totally avoided..

File: spec/models/course/assessment/category_spec.rb (hunk 5 above)
Comment (lowjoel) [inline spec/models/course/assessment/category_spec.rb]: But there's half-width (0) and full-width （0）

File: app/models/course/material/folder.rb (hunk 4 above)
Comment (allenwq) [inline app/models/course/material/folder.rb]: :+1:  Yep, agree !

File: spec/models/course/assessment/category_spec.rb (hunk 5 above)
Comment (angathorion) [inline spec/models/course/assessment/category_spec.rb]: They use half-width in Japan too for these problems.

Comment (lowjoel) [inline spec/models/course/assessment/category_spec.rb]: So the half-width space, and the half-width parentheses, with half-width numerals?

Comment (angathorion) [inline spec/models/course/assessment/category_spec.rb]: Yeah. That is the case.

File: app/models/course/material/folder.rb (hunk 3 above)
Comment (allenwq) [inline app/models/course/material/folder.rb]: I think it's necessary ? it's only saved once.

Suppose `owner` changes its name to `"AA"`, then folder name will also be changed to `"AA"` in owner's callback. But if `"AA"` is already taken and the folder is not valid, then folder cannot be saved. 

This function helps to update `"AA"` to `"AA(0)"` in this case.

Comment (allenwq) [inline app/models/course/material/folder.rb]: Do you have better suggestions ?  or you think that owner should handle this case it self ?

Comment (lowjoel) [inline app/models/course/material/folder.rb]: But your owner is calling next_valid_name. So I suggest you keep this callback, then make next_valid_name private.

File: spec/models/course/assessment/category_spec.rb (hunk 5 above)
Comment (allenwq) [inline spec/models/course/assessment/category_spec.rb]: That looks weird and it's hard to define the "correct" behaviour.  Suppose I uploaded a file named `file.txt` which its name is in english, but my language is chinese. Should the file be renamed to `file（０）.txt` ?

File: app/models/course/material/folder.rb (hunk 3 above)
Comment (allenwq) [inline app/models/course/material/folder.rb]: Oh, sure. my bad, that line is not needed any more.

**Comment by allenwq** [comment]
@lowjoel Please review again ! thanks !

**Comment by lowjoel** [comment]
@allenwq travis failed

File: app/models/course/material/folder.rb [hunk 6]
```diff
@@ -3,6 +3,8 @@ class Course::Material::Folder < ActiveRecord::Base
   include Course::ModelComponentHost::Component
 
   after_initialize :set_defaults, if: :new_record?
+  before_validation :assign_valid_name
+  before_save :assign_valid_name
```
Comment (lowjoel) [inline app/models/course/material/folder.rb]: Any circumstance where you need to run the callback in both before_validation and before_safe?

Comment (allenwq) [inline app/models/course/material/folder.rb]: Yes, like this case: https://github.com/Coursemology/coursemology2/blob/master/app/models/course/assessment/category.rb#L16

If you build a folder in `before_create`, and save the parent, then folder's `before_validation` callback won't be called.

Comment (lowjoel) [inline app/models/course/material/folder.rb]: then that's a problem with using before_create, and before_validation should be used instead?

File: app/models/course/material/folder.rb [hunk 7]
```diff
@@ -3,6 +3,7 @@ class Course::Material::Folder < ActiveRecord::Base
   include Course::ModelComponentHost::Component
 
   after_initialize :set_defaults, if: :new_record?
+  before_validation :assign_valid_name
```
Comment (allenwq) [inline app/models/course/material/folder.rb]: @lowjoel  I also rebased here, the before_save callback was removed.

Comment (lowjoel) [inline app/models/course/material/folder.rb]: meaning?

File: app/models/course/material/folder.rb (hunk 6 above)
Comment (allenwq) [inline app/models/course/material/folder.rb]: @lowjoel You might have forgot the context here

File: app/models/course/material/folder.rb (hunk 7 above)
Comment (lowjoel) [inline app/models/course/material/folder.rb]: so actually the before_save is not needed, right?

Comment (allenwq) [inline app/models/course/material/folder.rb]: Yep, it's not needed now, since we use after_initialize to build the folder

Comment (lowjoel) [inline app/models/course/material/folder.rb]: Awesome! The best code is code that doesn't need to be there :stuck_out_tongue:

Comment (allenwq) [inline app/models/course/material/folder.rb]: :-P Thanks !
